Error processing run files on first startup

On first startup on a new machine:

deploy-tools logs:
```
05:18:28.944 INFO  - Enabled AWS support.
05:18:28.965 INFO  - Enabled http authentication.
05:18:29.643 INFO  - ClusterManager: Instantiated module.
05:18:29.647 ERROR - Error during cleanup of run files Cannot read directory /opt/sw/vertx/current/run. Does not exist
```

application logs:
```
05:18:30.249 ERROR - Unhandled exception
rx.exceptions.OnErrorNotImplementedException: Cannot read directory /opt/sw/vertx/current/run. Does not exist
        at rx.Observable$26.onError(Observable.java:7493) ~[rxjava-1.0.14.jar:1.0.14]
```

```
Caused by: io.vertx.core.file.FileSystemException: Cannot read directory /opt/sw/vertx/current/run. Does not exist
        at io.vertx.core.file.impl.FileSystemImpl$11.perform(FileSystemImpl.java:666) ~[vertx-core-3.2.0.jar:na]
```
Possible fixes:
1. Have the infrastructure make sure that that folder exists and is writable by vert.x
2. Let the deploy-tools create the folder

--- a/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/AwsDeployApplication.java
+++ b/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/AwsDeployApplication.java
@@ -29,15 +29,17 @@ public class AwsDeployApplication extends AbstractVerticle {
         final DeployArtifactService deployArtifactService = new DeployArtifactService(getVertx(), deployconfig);
         final DeployConfigService deployConfigService = new DeployConfigService(getVertx(), deployconfig);
         final DefaultDeployService defaultDeployService = new DefaultDeployService(deployApplicationService, deployArtifactService, deployConfigService);
-        final AutoDiscoverDeployService autoDiscoverDeployService = new AutoDiscoverDeployService(deployconfig, defaultDeployService, getVertx());
-
-        deployApplicationService.cleanup().subscribe();
 
         this.createRunDir(deployconfig);
+
+        deployApplicationService.cleanup().subscribe();
         AwsService awsService = null;
+        AutoDiscoverDeployService autoDiscoverDeployService = null;
 
         if (deployconfig.isAwsEnabled()) {
             awsService = (new AwsService(getVertx(), deployconfig));
+            autoDiscoverDeployService = new AutoDiscoverDeployService(deployconfig, defaultDeployService, getVertx());
+
         }
 
         Router router = Router.router(getVertx());
@@ -65,13 +67,15 @@ public class AwsDeployApplication extends AbstractVerticle {
         server.listen(deployconfig.getHttpPort());
         initiated = true;
         LOG.info("{}: Instantiated module.", LogConstants.CLUSTER_MANAGER);
-        if (deployconfig.isAwsEnabled() && deployconfig.isAwsAutoDiscover()) {
+
+        if (deployconfig.isAwsEnabled() && deployconfig.isAwsAutoDiscover() && autoDiscoverDeployService != null) {
             autoDiscoverDeployService.autoDiscoverFirstDeploy();
         }
     }
 
     private void createRunDir(DeployConfig deployconfig) {
         if (!vertx.fileSystem().existsBlocking(deployconfig.getRunDir())) {
+            LOG.debug("Creating runDir '{}'.", deployconfig.getRunDir());
             vertx.fileSystem().mkdirBlocking(deployconfig.getRunDir());
         }
     }

--- a/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/DeployConfig.java
+++ b/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/DeployConfig.java
@@ -183,8 +183,8 @@ public class DeployConfig {
     }
 
     private DeployConfig withHttpAuth(JsonObject config) {
-        this.httpAuthUser = validateField(HTTP_AUTH_USER, config);
-        this.httpAuthPassword = validateField(HTTP_AUTH_PASSWORD, config);
+        this.httpAuthUser = validateField(HTTP_AUTH_USER, config, "");
+        this.httpAuthPassword = validateField(HTTP_AUTH_PASSWORD, config, "");
 
         if (!httpAuthUser.isEmpty() && !httpAuthPassword.isEmpty()) {
             LOG.info("Enabled http authentication.");

--- a/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/service/DeployApplicationService.java
+++ b/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/service/DeployApplicationService.java
@@ -114,7 +114,7 @@ public class DeployApplicationService implements DeployService<DeployApplication
                 .flatMap(x -> just(deployRequest));
     }
 
-    public Observable<?> cleanup() {
+    public Observable<Boolean> cleanup() {
         List<String> runningApplications = new ProcessUtils(config).listModules();
         FileSystem fs = new io.vertx.rxjava.core.Vertx(vertx).fileSystem();
 
@@ -125,8 +125,11 @@ public class DeployApplicationService implements DeployService<DeployApplication
                 .filter(s -> !s.isEmpty() && !runningApplications.contains(s))
                 .flatMap(file -> fs.deleteObservable(config.getRunDir() + file))
                 .toList()
-                .flatMap(x -> just(null))
-                .doOnError(t -> LOG.error("Error during cleanup of run files {}", t.getMessage()));
+                .flatMap(x -> just(Boolean.TRUE).doOnError(t -> LOG.error("error")))
+                .onErrorReturn(x -> {
+                    LOG.error("Error during cleanup of run files {}", x.getMessage());
+                    return Boolean.FALSE;
+                });
     }
 
     public void addApplicationDeployResult(boolean succeeded, String message, String deploymentId) {

--- a/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/service/DeployApplicationService.java
+++ b/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/service/DeployApplicationService.java
@@ -118,6 +118,7 @@ public class DeployApplicationService implements DeployService<DeployApplication
         List<String> runningApplications = new ProcessUtils(config).listModules();
         FileSystem fs = new io.vertx.rxjava.core.Vertx(vertx).fileSystem();
 
+
         return fs.readDirObservable(config.getRunDir())
                 .flatMapIterable(x -> x)
                 .flatMap(s -> just(Pattern.compile("/").splitAsStream(s).reduce((a, b) -> b).orElse("")))

--- a/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/util/ProcessUtils.java
+++ b/vertx-deploy/src/main/java/nl/jpoint/vertx/mod/deploy/util/ProcessUtils.java
@@ -16,7 +16,6 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
-import java.util.function.Function;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -45,8 +44,8 @@ public class ProcessUtils {
         List<String> moduleIds = this.listModules();
         return moduleIds.stream()
                 .map(this::parseModuleString)
-                .collect(Collectors.toMap((Function<JsonObject, String>) jsonObject -> jsonObject.getString(Constants.MAVEN_ID),
-                        (Function<JsonObject, String>) jsonObject -> jsonObject.getString(Constants.MODULE_VERSION)));
+                .collect(Collectors.toMap(jsonObject -> jsonObject.getString(Constants.MAVEN_ID),
+                        jsonObject -> jsonObject.getString(Constants.MODULE_VERSION)));
     }
 
     private JsonObject parseModuleString(String moduleString) {

